### DISCUSSION TITLE
Return an object from a function of a class without any output to the screen

### DISCUSSION THREAD

**Discussion by netcorefan1** [PowerShell/PowerShell#19686] [Q&A] [answered]
Hello,
I'm trying to apply the fluent builder pattern to a PWSH class. It works fine, but after spending hours I didn't managed to find a way to suppress the output printed to the screen.
```
$Global:MC = [MyClass ]::new()
MyClass  {
    [MyClass] Write([string]$msg) { 
        #... processing code ...
        return $this
    }
}
```
`$MC.Write('message 1').Write('Message 2')` print `MyClass`  to the screen  (and even all of its members if they're not hidden) which is something I don't want.
I can suppress with `return $this | Out-Null` , but this also return a `$null` object which breaks all subsequent chained calls.
I can also suppress with `$null = $MC.Write('message 1').Write('Message 2')` or `$MC.Write('message 1').Write('Message 2')  | Out-Null`, but this is also something that I don't want.

How can I just use the fluent builder pattern in a clean way like my first sample without having the screen messed up with unwanted output?

**Comment by rhubarb-geek-nz** [comment]
What happens is that results from functions or commands get written to the output pipeline if they have not be consumed by something else, eg assigning to a variable.

Then the final act in the pipeline in PowerShell is writing the contents of the output pipeline to the process output (eg stdout )

This is all consistent and by design.

If you don't want a value to go into the output pipeline then do

```
$null = $myThing.DoThis('fee').DoThat('fi').DoTheOther('fo').DontForget('fum')
```

**Comment by rhubarb-geek-nz** [comment, marked as answer]
You could also have an EndWrite method which returns [void]

```
class MyClass {
    [MyClass] Write([string]$msg) { 
        Write-Host $msg
        return $this
    }
    [void] EndWrite() {
    }
}

$obj = [MyClass]::new()

$obj.Write('foo').EndWrite()
```

**Comment by rhubarb-geek-nz** [comment]
Isn't the purpose of the builder pattern for the last called to be to `Build() `which will return the object that was built? 

If you return void then you break the builder pattern entirely. 

The point of the fluent builder pattern is that you use each response from the previous command in order to configure the next property. 

So every response is significant and needs to be used until the final `Build()`, and then you get the object for which the whole exercise was for.

**Comment by netcorefan1** [comment]
@rhubarb-geek-nz Thanks for your detailed response. Calling a final `EndWrite()` is still better than assigning to a $null variable, but is something that I'm trying to avoid because it is basically going to add pointless complexity.

You're right, the build pattern should always end up with a `Build()` method that returns the built object, but in my case is different. I'm not going to build an object, but I just run some actions that print some stuff on the screen (`Write-Host $msg` in your sample perfectly shows my use case), so  adding a `Finalize()` void just to suppress the screen output seems more than an hack. There is no way to enforce its usage, so if someone forget it, the screen will get messed up.

I decided to go with such pattern after having seen how much would have been easy to print stuff through chained calls and with very short amount of code, but I would have never imagined that I would have ended up with being forced to abort my idea for what seemed to be a very simple step like suppressing screen output. There must be a way to directly deal with objects and keep the screen silenced during operations.

**Comment by rhubarb-geek-nz** [comment]
I don't really think that the chained calls is the PowerShell way of doing things. It would use pipelines etc to achieve the same.

For example

```
PS> @("one","two","three") | Write-Host
one
two
three
PS> Write-Host "one" "two" "three"       
one two three
PS> @("one","two","three") | ForEach-Object { Write-Host $_ }
one
two
three
```

So rather than chaining calls I would suggest try piping all your values through Write-Host or another function

**Reply by netcorefan1** [reply]
Yes, that was how I was doing (`function Print([Parameter(ValueFromPipeline, ValueFromPipelineByPropertyName)] ...`). Chaining calls is still easier and achieve the goal with shorter code.

What about overriding `out-default` ?
I got it working from terminal:
```
PS home> function out-default {$input | out-null}
PS home> $MC.Write('message 1 and').Write('Message 2')
```
Output: `message 1 and Message 2`

However, in VSCODE debug mode the class name is still printed if I run the same instructions inside the script. Same if I run the script from the terminal unless I first declare the function like this: `PS home> function out-default {$input | out-null};  ./MyScript.ps1`
If it works from the terminal it should work directly from the script too.
There is a `Remove-WriteHost` function posted [here](https://stackoverflow.com/questions/14470953/powershell-how-to-capture-or-suppress-write-host) which seems to temporarely override and restore WriteHost function. Pheraps could be adapted to work with `Out-Default` once I understand why the override only work from terminal.

**Reply by rhubarb-geek-nz** [reply]
> Chaining calls is still easier and achieve the goal with shorter code.

Or perhaps not. If feels like you are going against the grain. I would suggest that if you are having to look for tricks and overrides to get it to work then you are actually building in technical debt rather than solving problems.

My general rule is if the programming is easy, you are probably working with the system, keep going. If the programming is hard then you are going against the system and there may be better ways to solve the problem. That applies to any SDK, language or framework.

**Comment by ThomasNieto** [comment]
Since the builder paradigm returns the object to the pipeline you'll need to redirect it, save it to a variable, or cast it to `[void]` if you don't want it printed to the screen.

**Comment by netcorefan1** [comment]
Thanks guys for your help. What I'm trying to do seems possible, but I am not convinced at all.

Create an exact copy of the command and wrap the content in `function out-default { }`:
```
$MetaData = New-Object System.Management.Automation.CommandMetaData (Get-Command out-default)
[System.Management.Automation.ProxyCommand]::Create($MetaData)  |  Out-File 'out-default.psm1'
```
Process the pipeline only if the value does not match the object we want to exclude:
```
process {
      try {
        if ($_ -notmatch "MyClass"){
          $steppablePipeline.Process($_)
        }      
      }
      catch {
        throw
      }
    }
```
Create a manifest which will autoload the module (this is to avoid dependency from a PS profile):
`RequiredModules = @("OutDefault")`

Now, from a terminal:
```
# first time:
PS home> ./MyScript.ps1
PS home> message 1 and Message 2 MyClass
# Second and subsequent times:
PS home> ./MyScript.ps1
PS home> message 1 and Message 2
```
I don't know why it only works after second run, but I have read somewhere in the issues section about this problem and I'm not sure if this is a bug and about possible workarounds. VSCODE ignores my override and always displays `MyClass` , but I suppose this may have to do on how VS hooks the shell (may be something I could fix with the appropriate settings).

I'm not sure if I should continue in this way and try to fix the issue or add an `EndWrite()` returning a `[void]` as suggested from @rhubarb-geek-nz  and @ThomasNieto . Initially I was oriented with the last option, but now that I've nearly achieved what was my initial goal I'm in doubt.

> My general rule is if the programming is easy, you are probably working with the system, keep going. If the programming is hard then you are going against the system and there may be better ways to solve the problem. That applies to any SDK, language or framework.

Yes, I feel exactly like I'm going to hack the system. Overrides, modules here and there. All of this just to suppress the output on the screen for a target object. I feel like, soon or latter, these tricks will have their own side effects. Unless there is a way to obtain the same result in a non intrusive way like this, maybe I'm better off with `EndWrite()` although, an empty `[void] EndWrite() {  }`, also sounds like a cheat. Any idea on how I could improve this?

What really surprises me is how is possible that an advanced and powerful shell like PWSH does not provide a way to handle such scenarios. Everything goes to the screen and this is good, but at least the user should be able to decide what should go and what should not go while still being able to access and process the filtered object within the code.

**Reply by rhubarb-geek-nz** [reply]
> All of this just to suppress the output on the screen for a target object.

It isn't the target object that is being written to the output pipeline, but the result of the call; which in the case of your programming style turns out to also be the target object.

There is a difference between Write-Output and Write-Host, however at the top-most level of a script those may appear to be the same, but once things are pipelined you will see the difference.

**Reply by rhubarb-geek-nz** [reply]
> What really surprises me is how is possible that an advanced and powerful shell like PWSH does not provide a way to handle such scenarios.

It does, and you mentioned it in your first post. You assign the result to "$null".

Think of it this way; in PowerShell nothing is thrown away. Either you use it or else it gets written to the output pipeline. If you don't want it to go to the output pipeline then you must use it, even if it is just by assigning it to the "$null" variable.

This even includes creating a directory

```
$null = New-Item -ItemType Directory -Path $ModulePath
```

If you don't assign that to $null then the newly created directory is written to the output pipeline. However if you wanted to do something with the newly created directory it had returned the object you can use to refer to it.

**Reply by ThomasNieto** [reply]
Another alternative is to use a PowerShell formatter for your custom type so when the object reaches the screen the formatter is applied to your object. You can define those by using a [format file](https://learn.microsoft.com/en-us/powershell/scripting/developer/format/formatting-file-overview). The module [EZOut](https://www.powershellgallery.com/packages/EZOut) simplifies creating the format file.

**Reply by netcorefan1** [reply]
> > What really surprises me is how is possible that an advanced and powerful shell like PWSH does not provide a way to handle such scenarios.
> 
> It does, and you mentioned it in your first post. You assign the result to "$null".
> 
> Think of it this way; in PowerShell nothing is thrown away. Either you use it or else it gets written to the output pipeline. If you don't want it to go to the output pipeline then you must use it, even if it is just by assigning it to the "$null" variable.
> 
> This even includes creating a directory
> 
> ```
> $null = New-Item -ItemType Directory -Path $ModulePath
> ```
> 
> If you don't assign that to $null then the newly created directory is written to the output pipeline. However if you wanted to do something with the newly created directory it had returned the object you can use to refer to it.

I perfectly understand. I come from C# and the MVVM pattern where UI and objects are two separate layers. In powershell these two layers are bound together and trying to apply the same principle is very hard because this is not how it has been designed.



> Another alternative is to use a PowerShell formatter for your custom type so when the object reaches the screen the formatter is applied to your object. You can define those by using a [format file](https://learn.microsoft.com/en-us/powershell/scripting/developer/format/formatting-file-overview). The module [EZOut](https://www.powershellgallery.com/packages/EZOut) simplifies creating the format file.

Thanks @ThomasNieto ! I will try your suggestion as my last attempt. Could you give me some tip on how EZOut should be applied in my sample code?
```
MyClass  {
    [MyClass] Write([string]$msg) { 
        #... processing code ...
        return $this
    }
}
```

**Reply by rhubarb-geek-nz** [reply]
> I perfectly understand. I come from C# and the MVVM pattern where UI and objects are two separate layers. In powershell these two layers are bound together and trying to apply the same principle is very hard because this is not how it has been designed.

So my recommendation is don't. Don't go against the grain, use PowerShell for what it is good at and use C# for what it is good at.
If you want to use C# programming techniques then write a C# assembly or C# Cmdlet. I don't think your are doing anybody including yourself any favours trying to work against the system.

**Reply by netcorefan1** [reply]
Thanks @rhubarb-geek-nz . I have looked at EZOut  and unfortunately it sill needs something like `$MC.Write('message 1 and').Write('Message 2') | EzNullView`. The best suggestion is the [one](https://github.com/orgs/PowerShell/discussions/19686#discussioncomment-5959489) suggested from @rhubarb-geek-nz . I need to share code between a C# assembly and PWSH, so your suggestion to move most of the code in C# (for which I'm more comfortable) is also goodI

Thanks
